src/lib.rs

```diff
@@ -142,112 +127,120 @@ impl Metadata {
     }
 }
 
+key_integer!(usize);
+
 impl Key for &str {
-    fn find<M: Miniconf>(self) -> core::option::Option<usize> {
+    #[inline]
+    fn find<M: Miniconf>(&self) -> core::option::Option<usize> {
         M::name_to_index(self)
     }
 }
 
-/// Trait exposing serialization/deserialization of elements by path and traversal by path/indices.
+/// Trait exposing serialization/deserialization of nodes by keys/paths and traversal.
 pub trait Miniconf {
-    /// Convert a name key to an index.
-    fn name_to_index(value: &str) -> core::option::Option<usize>;
+    /// Convert a node name to a node index.
+    fn name_to_index(name: &str) -> core::option::Option<usize>;
 
-    /// Deserialize an element by key.
+    /// Serialize a node by keys.
     ///
     /// # Args
-    /// * `keys`: An `Iterator` identifying the element. The iterator items
-    ///    must support conversion to graph indices through [`Key`]
-    /// * `de`: A `Deserializer` to deserialize the value.
+    /// * `keys`: An `Iterator` of `Key`s identifying the node.
+    /// * `ser`: A `Serializer` to to serialize the value.
     ///
     /// # Returns
-    /// [`Ok`] on success, [Error] on failure.
-    fn set_by_key<'a, K, D>(&mut self, keys: K, de: D) -> Result<D::Error>
+    /// Node depth on success.
+    fn get_by_key<K, S>(&self, keys: K, ser: S) -> Result<usize, Error<S::Error>>
     where
         K: Iterator,
         K::Item: Key,
-        D: serde::Deserializer<'a>;
+        S: serde::Serializer;
 
-    /// Serialize an element by key.
+    /// Deserialize an node by keys.
     ///
     /// # Args
-    /// * `keys`: An `Iterator` identifying the element. The iterator items
-    ///    must support conversion to graph indices through [`Key`]
-    /// * `ser`: A `Serializer` to to serialize the value.
+    /// * `keys`: An `Iterator` of `Key`s identifying the node.
+    /// * `de`: A `Deserializer` to deserialize the value.
     ///
     /// # Returns
-    /// [`Ok`] on success, [Error] on failure.
-    fn get_by_key<K, S>(&self, keys: K, ser: S) -> Result<S::Error>
+    /// Node depth on success
+    fn set_by_key<'a, K, D>(&mut self, keys: K, de: D) -> Result<usize, Error<D::Error>>
     where
         K: Iterator,
         K::Item: Key,
-        S: serde::Serializer;
+        D: serde::Deserializer<'a>;
 
-    /// Call `func` for each element on the path described by a key.
+    /// Call a function for each node on the path described by keys.
     ///
     /// Traversal is aborted once `func` returns an `Err(E)`.
     ///
-    /// May not exhaust the iterator if a leaf is found early. i.e. keys may be too long.
-    /// If `Self` is a leaf, nothing will be consumed from the iterator
-    /// and [`Ok::Leaf(0)`] will be returned.
-    /// If `Self` is non-leaf (internal) and the iterator is exhausted (empty),
-    /// [`Ok::Internal(0)`] will be returned.
+    /// May not exhaust `keys` if a leaf is found early. i.e. `keys`
+    /// may be longer than required.
```

Comment:
What's the design decision driving this? Seems odd to just ignore extra keys in the iterator when that would imply that we're traversing to something potentially not specified by the user.

Take for example one firmware version exposes the path: "/settings/enable"

Then, in a later version, firmware exposes "/settings/enable/power_disable".

If the user wrong to `/settings/enable/power_disable` on FW V1, would that mean that it resolves to "/settings/enable"?

Reply:
It seems this isn't allowed for `get/set_by_key`, so I'm confused why it's the case here.

Reply:
This is in line with the previous behaviour where the iterator state was allowed to be too long.
The alternative would not change much to the method: you'd get `Err(TooLong(depth))` instead of `Ok(depth)` back but the call pattern to `func()` would always be the same.
And it would make the match in `PathIter::next()` a bit more complicated (as we don't want to find the right depth for each `path()` call first).
The important thing is probably the sideeffect:
Not checking for too long paths here also allows the remaining path elements to be used for something else since the iterator can be passed by mutable reference and isn't consumed, (maybe like URLs?).

For `{get,set}_by_key` it's decidedly different. There I want the keys to be unambiguous and they can only ever meaningfully refer to a leaf and want to exhaust the iterator.